include/nbl/system/IAsyncQueueDispatcher.h

```diff
@@ -16,9 +17,22 @@ namespace impl
     public:
         struct request_base_t
         {
+            std::mutex mtx;
             // wait on this for result to be ready
             std::condition_variable cvar;
-            bool ready = false;
+            std::atomic_bool ready = false;
+
+            std::unique_lock<std::mutex> lock()
+            {
+                return std::unique_lock<std::mutex>(mtx);
+            }
+
+            std::unique_lock<std::mutex> wait()
+            {
+                std::unique_lock<std::mutex> lk(mtx);
+                cvar.wait(lk, [this]() { return this->ready.load(); });
+                return lk;
```

Comment:
is copy ellision guaranteed?

because otherwise you'll end up with dtor, copy ctor (unlock in the middle)

Reply:
it's a move here or maybe copy ellision, but im almost sure its not guaranteed in this case

Reply:
so it might result in a pointless lock&unlock before it actually locks

Reply:
no, move doesnt touch mutex state at all

Reply:
and theres no copy ctor/assignment

Reply:
https://en.cppreference.com/w/cpp/thread/unique_lock/unique_lock

Reply:
> 
> 
> no, move doesnt touch mutex state at all

ok thanks